### ISSUE TITLE
host detection/`NEXTAUTH_URL` breaks in some cases

### ISSUE THREAD

**Issue by mchataigner** [nextauthjs/next-auth#5953]
### Provider type

Google

### Environment

  System:
    OS: Linux 6.0 Arch Linux
    CPU: (16) x64 12th Gen Intel(R) Core(TM) i7-1260P
    Memory: 20.61 GB / 31.06 GB
    Container: Yes
    Shell: 5.9 - /usr/bin/zsh
  Binaries:
    Node: 16.18.1 - ~/.nvm/versions/node/v16.18.1/bin/node
    npm: 8.19.2 - ~/.nvm/versions/node/v16.18.1/bin/npm
  Browsers:
    Chromium: 108.0.5359.94
    Firefox: 107.0.1
  npmPackages:
    next: 13.0.2 => 13.0.2
    next-auth: ^4.18.0 => 4.18.0
    react: 18.2.0 => 18.2.0

### Reproduction URL

https://next-auth-google-issue-repro.vercel.app/
https://github.com/mchataigner/next_auth_google_issue_repro

### Describe the issue

I have deployed on Vercel a simple app with next-auth and Google Oauth provider.
It works well locally but I get an error when trying to login in production.
It seems the error is coming from /api/auth/providers returning an empty object despite having multiple providers set in my configuration:
* Discord alone => works fine
* Google alone => error (empty providers object)
* Google + Discord => error (empty providers object)

I tried with next-auth 4.15.1 and 4.18.0 with no success
![screenshot20221205_102909](https://user-images.githubusercontent.com/2926444/205602349-a2c13821-170b-4584-b3c8-f02ce2159744.png)
Trying again this morning, and it worked a couple of times and then stopped working again.
It seems to work on cold start but not on hot ones.

### How to reproduce

Create simple app and add Google Oauth provider.

### Expected behavior

Adding Google provider should expose google login for the application or at least avoid crashing other providers.
![screenshot20221205_101539](https://user-images.githubusercontent.com/2926444/205600459-a7df61f5-ca57-4430-8a37-88366d92eac0.png)

**Comment by llegomark** [comment]
I have the same problem. 

It works fine on my local host (Discord and Google), but on production, it crashes.

**Comment by mchataigner** [comment]
hello @balazsorban44 , I've seen that you added the incomplete tag on this issue.
Could you please tell me what is missing?

**Comment by balazsorban44** [comment]
> Create simple app and add Google Oauth provider.

This is unfortunately not a reproduction. Please add a link to your public repository, preferably a minimal one, that only focuses on reproducing the issue, and removing any unnecessary code. Will need more info on the logs you are getting in production as well to guide us on what could be the problem.

**Comment by MJomaa** [comment]
The affected version is 4.18.0.

It is a bug introduced here which empties your social providers list:
https://github.com/nextauthjs/next-auth/pull/5792/files

If you use unstable_getServerSession() and call getProviders() you can reproduce it.

A fix is on the way:
https://github.com/nextauthjs/next-auth/pull/5973/files

For now downgrade to 4.17.0.

**Comment by askkaz** [comment]
Amazing, thanks for the update! This is our exact problem :-)

**Comment by mchataigner** [comment]
Awesome, thank you very much. I'm new to this library and did not know where to look.

**Comment by eshaffer321** [comment]
I had this generic error message too and was stumped for a while. Turns out I had 2 `[..nextauth].ts` files and there was no indication of that in any of the logs or with debug mode on. Posting this in case someone else gets hit by this.

**Comment by balazsorban44** [comment]
@mchataigner FWIW, your production app url is still not a reproduction. here, have a read at this for future reference https://github.com/nextauthjs/next-auth/blob/main/.github/actions/issue-validator/repro.md 🙏

**Comment by mchataigner** [comment]
sure @balazsorban44, I was about to publish a sample app on public github and vercel but as the root cause is already identified and a fix is already on the way I did not finish it.

**Comment by balazsorban44** [comment]
Should be out in `4.18.1`, please test it and let us know!

**Comment by mchataigner** [comment]
I've just tested out and it still seems not working but differently.
With 4.18.1, I get some errors "Invalid redirect URL" causing 500 errors on some authenticated endpoints.

Reverting to 4.17.0 solves the issue.

I published a demo app here: https://next-auth-google-issue-repro.vercel.app/
with source code available here: https://github.com/mchataigner/next_auth_google_issue_repro

**Comment by balazsorban44** [comment]
Seems like something is off with your deployment, every `next-auth` route returns 400, eg: https://next-auth-google-issue-repro.vercel.app/api/auth/session

Make sure you follow: https://next-auth.js.org/deployment#vercel

Locally everything works fine for me. :thinking:

**Comment by balazsorban44** [comment]
Was a bit eager to push the first patch and did not properly test it before publishing, but `4.18.3` should fix it! Let me know!

**Comment by balazsorban44** [comment]
Working, deployed example here: https://next-auth-example.vercel.app/

**Comment by balazsorban44** [comment]
It looks like even that was not enough to fix it unfortunately,

**Comment by balazsorban44** [comment]
@mchataigner I think I got it fixed, could you test out `pnpm add next-auth@0.0.0-pr.5991.5c72424c` and report back if that release works for you? If so, #5991 is ready to be merged/released. :pray:

**Comment by innellea** [comment]
> @mchataigner I think I got it fixed, could you test out `pnpm add next-auth@0.0.0-pr.5991.5c72424c` and report back if that release works for you?
> 
> I tested it on https://next-auth-5953.vercel.app
> 
> If so, #5991 is ready to be merged/released. 🙏

@balazsorban44  I tested with GoogleProvider, and it works flawlessly, just as before!

**Comment by chrisflatley** [comment]
Different issue to the OP, but relates to the code from `next-auth@0.0.0-pr.5991.5c72424c`.
As of v4.18.2, changing the basePath doesn't work (i.e. NEXTAUTH_URL=http://localhost:3000/CUSTOM/api/auth)

```
# if
NEXTAUTH_URL=http://localhost:3000/custom/api/auth

# Then getUrl() returns
"http://localhost:3000/custom/api/auth" + url
# For example:
http://localhost:3000/custom/api/auth/api/auth/error
```

I also get `Error: This action with HTTP GET is not supported by NextAuth.js` if I visit say http://localhost:3000/custom/api/auth/session or signin.

It might be my setup (v4.18.1 works) so I'll investigate  more and create a repro & issues if necessary,.

I thought I'd comment here before you merge.

**Comment by balazsorban44** [comment]
Hmm. Yeah, that seems to be related, I'll look into it before merging, thanks for reporting!

**Comment by AndreyMay** [comment]
After upgrading to v4.18.3 started seeing the following issue with Google signin when deployed.
Downgrade to v4.17.0 helped.

```
[next-auth][error][OAUTH_CALLBACK_ERROR] 
https://next-auth.js.org/errors#oauth_callback_error checks.state argument is missing {
  error: TypeError: checks.state argument is missing
      at Client.callback (/var/task/node_modules/openid-client/lib/client.js:387:13)
      at oAuthCallback (/var/task/apps/app/node_modules/next-auth/core/lib/oauth/callback.js:127:29)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async Object.callback (/var/task/apps/app/node_modules/next-auth/core/routes/callback.js:52:11)
      at async AuthHandlerInternal (/var/task/apps/app/node_modules/next-auth/core/index.js:167:28)
      at async AuthHandler (/var/task/apps/app/node_modules/next-auth/core/index.js:309:28)
      at async NextAuthHandler (/var/task/apps/app/node_modules/next-auth/next/index.js:24:20)
      at async /var/task/node_modules/@sentry/nextjs/cjs/config/wrappers/withSentryAPI.js:132:31
      at async Object.apiResolver (/var/task/node_modules/next/dist/server/api-utils/node.js:366:9)
      at async NextNodeServer.runApi (/var/task/node_modules/next/dist/server/next-server.js:481:9) {
    name: 'OAuthCallbackError',
    code: undefined
  },
  providerId: 'google',
  message: 'checks.state argument is missing'
}
```

Separately, spotted a difference in `signIn` callback behavior when returning a relative URL. It worked fine before v4.17.0, but after v4.17.0 only the absolute url works.

**Comment by balazsorban44** [comment]
@AndreyMay could you test out https://github.com/nextauthjs/next-auth/issues/5953#issuecomment-1342923567? That includes a fix for cookie handling, which is likely the source of your issue.

**Comment by AndreyMay** [comment]
@balazsorban44 `next-auth@0.0.0-pr.5991.5c72424c` solved the `checks.state argument is missing` issue.

**Comment by balazsorban44** [comment]
Great to hear, that's out on `4.18.4` now. :+1:

Focusing on the host detection part, it will be in the next patch.

**Comment by chrisflatley** [comment]
@balazsorban44  for the host detection I think it would be nice it NEXTAUTH_URL was taken as the right value.
The current flow is if NEXTAUTH_URL but gets overridden  VERCEL or proxy.  

(Untested) but I feel the code should look like

```js
export function getURL(
  url: string | undefined | null,
  trusted: boolean | undefined,
  forwardedValue: string | string[] | undefined | null
): URL | Error {
  try {
      if (!url) throw new TypeError("Invalid URL, cannot determine action")
  
     let host

    if(process.env.NEXTAUTH_URL) {
         // TODO this is not technically just the host see below
         host = process.env.NEXTAUTH_URL
    } else if(process.env.AUTH_TRUST_HOST && forwardedValue) {
         host = Array.isArray(forwardedValue) ? forwardedValue[0] : forwardedValue
    } else if(process.env.VERCEL) {
         host = process.env.VERCEL
    } else if(process.env.NODE_ENV !== "production") {
        host =  "http://localhost:" + (process.env.PORT ?? "3000")
    } else {
        throw new TypeError("Invalid host")
    }

    // if we just have a domain, add a protocol
    if (!host.startsWith("http://") &&! !host.startsWith("https://")) {
      // NOTE if you don't want this to be HTTPS you should explicitly set your NEXTAUTH_URL
      host = `https://${host}`
    }
    
    return new URL(`https://${host}${url}`)
  } catch (error) {
    return error as Error
  }
}
```

This doesn't deal with the basePath issue, but I'm not sure if that's a bug here or with the more to use the Request/Response which happened in v4.18.2 (or perhaps both?!)

The basePath issue could be addressed like:

```js
    if(process.env.NEXTAUTH_URL) {
        const nextAuthUrl = new URL(process.env.NEXTAUTH_URL)
        host = nextAuthUrl.origin 
        const  basePath = nextAuthUrl.pathname
        url = url.replace("/api/auth", basePath) 
    }
```

**Comment by chrisflatley** [comment]
Re the basePath: the `utils/web.js` `toInternalRequest()` defines `action: nextauth[0]` (after a slice(3)), which with a base path might be /custom/api/auth/signin  is `auth` not `signin`.

**Comment by Dav3rs** [comment]
Every next-auth route returns 400, with code that worked the last week. It's madness.

**Comment by AndreyMay** [comment]
Looks like there are more breaking changes in v4.18.x, just discovered that signin (email and google) stopped working on mobile while working on desktop. I'm setting additional cookies in `signIn` event and somehow on mobile it stopped working. Downgraded to v4.17.0 again and it's back to normal.

**Comment by steven-tey** [comment]
Running into this issue as well with https://github.com/domeccleston/sharegpt (`next-auth : ^4.18.0`)

Works fine locally, even with `next build`.

Was able to fix it by defining a `NEXTAUTH_URL` env var in Vercel, but it was driving me crazy for half a day 😅

**Comment by balazsorban44** [comment]
This should be fixed in `4.18.5`, please give it a try!
